libs/partners/pinecone/langchain_pinecone/vectorstores.py

```diff
@@ -0,0 +1,484 @@
+                    "or `PINECONE_API_KEY` environment variable"
+                )
+
+            _index_name = index_name or os.environ.get("PINECONE_INDEX_NAME") or ""
+            if not _index_name:
+                raise ValueError(
+                    "Pinecone index name must be provided in either `index_name` "
+                    "or `PINECONE_INDEX_NAME` environment variable"
+                )
+
+            # needs
+            client = PineconeClient(api_key=_pinecone_api_key)
+            self._index = client.Index(_index_name)
+
+    @property
+    def embeddings(self) -> Optional[Embeddings]:
+        """Access the query embedding object if available."""
+        return self._embedding
+
+    def add_texts(
+        self,
+        texts: Iterable[str],
+        metadatas: Optional[List[dict]] = None,
+        ids: Optional[List[str]] = None,
+        namespace: Optional[str] = None,
+        batch_size: int = 32,
+        embedding_chunk_size: int = 1000,
+        **kwargs: Any,
+    ) -> List[str]:
+        """Run more texts through the embeddings and add to the vectorstore.
+
+        Upsert optimization is done by chunking the embeddings and upserting them.
+        This is done to avoid memory issues and optimize using HTTP based embeddings.
+        For OpenAI embeddings, use pool_threads>4 when constructing the pinecone.Index,
+        embedding_chunk_size>1000 and batch_size~64 for best performance.
+        Args:
+            texts: Iterable of strings to add to the vectorstore.
+            metadatas: Optional list of metadatas associated with the texts.
+            ids: Optional list of ids to associate with the texts.
+            namespace: Optional pinecone namespace to add the texts to.
+            batch_size: Batch size to use when adding the texts to the vectorstore.
+            embedding_chunk_size: Chunk size to use when embedding the texts.
+
+        Returns:
+            List of ids from adding the texts into the vectorstore.
+
+        """
+        if namespace is None:
+            namespace = self._namespace
+
+        texts = list(texts)
+        ids = ids or [str(uuid.uuid4()) for _ in texts]
+        metadatas = metadatas or [{} for _ in texts]
+        for metadata, text in zip(metadatas, texts):
+            metadata[self._text_key] = text
+
+        # For loops to avoid memory issues and optimize when using HTTP based embeddings
+        # The first loop runs the embeddings, it benefits when using OpenAI embeddings
+        # The second loops runs the pinecone upsert asynchronously.
+        for i in range(0, len(texts), embedding_chunk_size):
+            chunk_texts = texts[i : i + embedding_chunk_size]
+            chunk_ids = ids[i : i + embedding_chunk_size]
+            chunk_metadatas = metadatas[i : i + embedding_chunk_size]
+            embeddings = self._embedding.embed_documents(chunk_texts)
+            async_res = [
```

Comment:
Can you please support non-async? Pinecone supports it via `index.upsert(..., async_req=False)`. Here's an easy fix: https://github.com/langchain-ai/langchain/pull/16753